### PR TITLE
🐛 fix/network/rtb: delete rtb handling err when failed to create routes

### PR DESCRIPTION
**Pull request by mtulio**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

This fix creates an error handler when the route table reconcile loop returns error when creating routes. It will prevent left the route table behind.

Currently this issue is triggered when an invalid route entry is added, creating a new stale table every new reconciliation, growing to hundreds in a few minutes. This is one example triggering the issue (a subset of Carrier Gateway support adding in #4901):

~~~sh
E0403 01:04:42.329759 1157154 awscluster_controller.go:327] "failed to reconcile network" err=<
	failed to create route in route table "rtb-067f352c09371fe71": {
	  CarrierGatewayId: "cagw-0ac129c044b2cd1b6",
	  DestinationCidrBlock: "0.0.0.0/0"
	}: MissingParameter: The request must contain exactly one of gatewayId, natGatewayId, networkInterfaceId, vpcPeeringConnectionId, egressOnlyInternetGatewayId, transitGatewayId, localGatewayId, carrierGatewayId, vpcEndpointId, coreNetworkArn or instanceId
		status code: 400, request id: 570b91c9-e228-4ee9-a939-1f5fe0313165
~~~

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [ ] includes documentation (N/A)
- [x] includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)
- [x] adds unit tests
- [ ] adds or updates e2e tests (N/A)

**Release note**:

```release-note
Fix the error handling by deleting the created route table on errors when creating routes.
```

### CODE REVIEW

File: pkg/cloud/services/network/routetables.go [hunk 1]
```diff
@@ -302,8 +311,11 @@ func (s *Service) createRouteTableWithRoutes(routes []*ec2.Route, isPublic bool,
 			}
 			return true, nil
 		}, awserrors.RouteTableNotFound, awserrors.NATGatewayNotFound, awserrors.GatewayNotFound); err != nil {
-			// TODO(vincepri): cleanup the route table if this fails.
 			record.Warnf(s.scope.InfraCluster(), "FailedCreateRoute", "Failed to create route %s for RouteTable %q: %v", route.GoString(), *out.RouteTable.RouteTableId, err)
+			errDel := s.deleteRouteTable(out.RouteTable)
+			if errDel != nil {
+				s.scope.Trace("Failed to delete RouteTable %q: %v", *out.RouteTable.RouteTableId, errDel)
```
Comment (mtulio) [inline pkg/cloud/services/network/routetables.go:317]: Hey @vincepri  - is this a recommended approach. The idea here is not recording the route table delete failure as the `FailedCreateRoute` had been recorded. Is it make sense?

File: pkg/cloud/services/network/routetables.go [hunk 2]
```diff
@@ -214,6 +214,32 @@ func (s *Service) describeVpcRouteTablesBySubnet() (map[string]*ec2.RouteTable,
 	return res, nil
 }
 
+func (s *Service) deleteRouteTable(rt *ec2.RouteTable) error {
```
Comment (mtulio) [inline pkg/cloud/services/network/routetables.go:217]: there is nothing new in this function, just isolating from `deleteRouteTables()` to be reused in the error handling for create route.

**Comment by mtulio** [comment]
@vincepri @damdo would you mind to stamp labels to test?

File: pkg/cloud/services/network/routetables.go (hunk 1 above)
Comment (mtulio) [reply pkg/cloud/services/network/routetables.go:317]: + @damdo

**Comment by mtulio** [comment]
/assign @damdo

e2e passing. Unsure why GH checks failed.

I am happy to introduce units for those functions. Let me know if that approach is feasible before writing more code.

**Review by damdo — COMMENTED** [review]
Left a comment but the approach LGTM.
I think, yes, if you can add a unit test that covers this, it would be great.
Thanks!

**Comment by mtulio** [comment]
> Left a comment but the approach LGTM. I think, yes, if you can add a unit test that covers this, it would be great. Thanks!

@damdo unit tests added, and everything is passing now.

**Review by damdo — COMMENTED** [review]
Thanks @mtulio 
/lgtm

Assigning approvers
/assign @dlipovetsky @vincepri @richardcase @Ankitasw

**Review by vincepri — APPROVED** [review]
/approve

**Pull request merged**